src/ceph-volume/ceph_volume/util/disk.py

```diff
@@ -195,6 +195,26 @@ def device_family(device):
 
     return devices
 
+def is_rotational(device):
+    """
+    Returns true if a device is rotational or not.
+    """
+    labels = ['ROTA']
+    command = ['lsblk', '-P', '-p', '-o', ','.join(labels), device]
+    out, err, rc = process.call(command)
+    device = []
+    for line in out:
+        device.append(_lsblk_parser(line))
+
+    for i in device:
+        if type(i) is dict and "ROTA" in i.keys():
+            if i['ROTA'] == '1':
+                rotational = True
+            else:
+                rotational = False
+
+    return rotational
```

Comment:
It's been a while ago, but IIRC this will not work for LVM based devices, which is the default in ceph-volume (LVM). On the host the names can be:
/dev/mapper/osd.0-osd.0
/dev/dm-0
/dev/osd.0/osd.0

And those are treated in a different way than raw block devices. And passing one of these to disk.lsblk will not work for ROTA. But I will retest again with your suggestion and see if that is (still) true. If you can test yourself and see if that works for you or not ... might be the quickest way to see what goes wrong and where.

Reply:
```python
>>> disk.lsblk('/dev/mapper/vg_test-lv_test')
{'NAME': 'vg_test-lv_test', 'KNAME': 'dm-0', 'PKNAME': '', 'MAJ:MIN': '252:0', 'FSTYPE': '', 'MOUNTPOINT': '', 'LABEL': '', 'UUID': '', 'RO': '0', 'RM': '0', 'MODEL': '', 'SIZE': '100G', 'STATE': 'running', 'OWNER': 'ceph', 'GROUP': 'ceph', 'MODE': 'brw-rw----', 'ALIGNMENT': '0', 'PHY-SEC': '512', 'LOG-SEC': '512', 'ROTA': '1', 'SCHED': '', 'TYPE': 'lvm', 'DISC-ALN': '0', 'DISC-GRAN': '512B', 'DISC-MAX': '2G', 'DISC-ZERO': '0', 'PARTLABEL': ''}
```

```python
>>> disk.lsblk('/dev/mapper/vg_test-lv_test')['ROTA']
'1'
>>> disk.lsblk('/dev/dm-0')['ROTA']
'1'
>>> disk.lsblk('/dev/vg_test/lv_test')['ROTA']
'1'
>>>

Reply:
In a couple of places in the code the key "ROTA" defaults to "1" if it cannot be determined for whatever reason. So to make sure it really checks the state of the device (instead of returning a default "1")  I would like to ask you if you can test with a SSD and / or  NVMe disk. Or to similate a non-rotational device in a virtual machine.

For the VM option with libvirt you would would need to add an extra attribute to the domain xml: "rotation_rate" (and poweroff / poweron the VM afterwards):

(virsh edit domain would let you do that in an editor). Example:

```
 </disk>
     ...
     ...
      </source>
      <target dev='vdb' bus='scsi' rotation_rate='1'/>
      <alias name='scsi0-0-2-0'/>
      <address type='drive' controller='0' bus='0' target='2' unit='0'/>
 </disk>
```

Note:  in libvirt a rotation_rate of 1 indicates _non_rotational_ as opposed to what is used in the linux kernel ("0"). See [libvirt domain format](https://libvirt.org/formatdomain.html).

Note: merely doing a "echo 0 > /sys/block/vdb/queue/rotational" won't work, as I observed this value gets reset when ceph-volume runs.

Reply:
```console
# lsblk
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
vda     253:0    0   11G  0 disk
`-vda1  253:1    0   10G  0 part /
vdb     253:16   0  200G  0 disk
vdc     253:32   0  200G  0 disk
vdd     253:48   0  200G  0 disk
vde     253:64   0  200G  0 disk
vdf     253:80   0  200G  0 disk
vdg     253:96   0  200G  0 disk
vdh     253:112  0  200G  0 disk
vdi     253:128  0  200G  0 disk
vdj     253:144  0  200G  0 disk
vdk     253:160  0  200G  0 disk
nvme0n1 259:0    0  200G  0 disk
nvme1n1 259:1    0  200G  0 disk
nvme2n1 259:2    0  200G  0 disk
```
```python
>>> from ceph_volume.util.device import Device
>>> Device('/dev/nvme0n1').rotational
False
>>> Device('/dev/vdk').rotational
True
>>>
```

Reply:
Sorry, I should have been a bit more specific and ask for LVM logical volume test. See below:

```
TEST root@dennisosd2:~$ lsblk
NAME          MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda             8:0    0   10G  0 disk 
├─sda1          8:1    0   60M  0 part /boot/efi
└─sda2          8:2    0   10G  0 part 
  ├─LVM0-SWAP 253:1    0  976M  0 lvm  [SWAP]
  └─LVM0-ROOT 253:2    0    9G  0 lvm  /
sdb             8:16   0  100G  0 disk 
└─osd.4-osd.4 253:4    0  100G  0 lvm  
sdc             8:32   0  100G  0 disk 
└─osd.5-osd.5 253:3    0  100G  0 lvm  
vda           252:0    0  100G  0 disk 
└─osd.3-osd.3 253:0    0  100G  0 lvm
```

Those are all disks in a virtual machine. `/dev/sda` (OS), `/dev/sdb`, `/dev/vda` are rotational devices. `/dev/sdc` is a non-rotational device.  

### Rotational device tests:

```
TEST root@dennisosd2:~$ cat /sys/block/sdb/queue/rotational 
1
```
```
ls -lrt /dev/mapper/
total 0
crw------- 1 root root 10, 236 Nov  9 11:57 control
lrwxrwxrwx 1 root root       7 Nov  9 11:57 LVM0-ROOT -> ../dm-2
lrwxrwxrwx 1 root root       7 Nov  9 11:57 LVM0-SWAP -> ../dm-1
lrwxrwxrwx 1 root root       7 Nov  9 11:57 osd.4-osd.4 -> ../dm-4
lrwxrwxrwx 1 root root       7 Nov  9 11:57 osd.3-osd.3 -> ../dm-0
lrwxrwxrwx 1 root root       7 Nov  9 11:57 osd.5-osd.5 -> ../dm-3

cat /sys/block/dm-4/queue/rotational 
1

>>> from ceph_volume.util import disk
>>> disk.lsblk('/dev/sdb')['ROTA']
'1'
>>> disk.lsblk('/dev/osd.4/osd.4')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'
>>> disk.lsblk('/dev/dm-4')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'

>>> from ceph_volume.util.device import Device
>>> Device('/dev/sdb').rotational
Executable lvs not found on the host, will return lvs as-is
Executable pvs not found on the host, will return pvs as-is
Executable pvs not found on the host, will return pvs as-is
True
>>> Device('/dev/osd.4/osd.4').rotational
Executable lvs not found on the host, will return lvs as-is
True
>>> Device('/dev/dm-4').rotational
Executable lvs not found on the host, will return lvs as-is
True
```

### Non-rotational device tests

```
TEST root@dennisosd2:~$ cat /sys/block/sdc/queue/rotational 
0

TEST root@dennisosd2:~$ cat /sys/block/dm-3/queue/rotational 
0

>>> from ceph_volume.util import disk
>>> disk.lsblk('/dev/sdc')['ROTA']
'0'
>>> disk.lsblk('/dev/osd.5/osd.5')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'
>>> disk.lsblk('/dev/dm-5')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'


>>> from ceph_volume.util.device import Device
>>> Device('/dev/sdc').rotational
Executable lvs not found on the host, will return lvs as-is
Executable pvs not found on the host, will return pvs as-is
Executable pvs not found on the host, will return pvs as-is
False
```

Note that this is correct, it is indeed non-rotational

```
>>> Device('/dev/osd.5/osd.5').rotational
Executable lvs not found on the host, will return lvs as-is
True
```

Note that this is incorrect, it is the same device but now identified as a rotational device

```
>>> Device('/dev/dm-5').rotational
Executable lvs not found on the host, will return lvs as-is
 stderr: blkid: error: /dev/dm-5: No such file or directory
failed to determine if device /dev/dm-5 is BlueStore. device should not be used to avoid false negatives. err: [Errno 2] No such file or directory: '/dev/dm-5'
failed to determine if device /dev/dm-5 is BlueStore. device should not be used to avoid false negatives. err: [Errno 2] No such file or directory: '/dev/dm-5'
 stderr: Unknown device "/dev/dm-5": No such file or directory
True
```

Note that this is incorrect, it is the same device but now identified as a rotational device

Regarding logging lvs/pvs not found as an Executable ...

```
TEST root@dennisosd2:~$ which lvs
/usr/sbin/lvs
TEST root@dennisosd2:~$ /usr/sbin/lvs
  WARNING: PV /dev/sdc in VG osd.5 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/sdb in VG osd.4 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/vda in VG osd.3 is using an old PV header, modify the VG to update.
  LV    VG    Attr       LSize    Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  ROOT  LVM0  -wi-ao----    8.98g                                                    
  SWAP  LVM0  -wi-ao----  976.00m                                                    
  osd.3 osd.3 -wi-ao---- <100.00g                                                    
  osd.4 osd.4 -wi-ao---- <100.00g                                                    
  osd.5 osd.5 -wi-ao---- <100.00g

TEST root@dennisosd2:~$ which pvs
/usr/sbin/pvs
TEST root@dennisosd2:~$ /usr/sbin/pvs
  WARNING: PV /dev/sdc in VG osd.5 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/sdb in VG osd.4 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/vda in VG osd.3 is using an old PV header, modify the VG to update.
  PV         VG    Fmt  Attr PSize    PFree
  /dev/sda2  LVM0  lvm2 a--    <9.94g    0 
  /dev/sdb   osd.4 lvm2 a--  <100.00g    0 
  /dev/sdc   osd.5 lvm2 a--  <100.00g    0 
  /dev/vda   osd.3 lvm2 a--  <100.00g    0

TEST root@dennisosd2:~$ stat /usr/sbin/lvs
  File: /usr/sbin/lvs -> lvm
  Size: 3         	Blocks: 0          IO Block: 4096   symbolic link
Device: fd02h/64770d	Inode: 286530      Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-11-09 11:31:15.804000000 +0100
Modify: 2020-02-13 22:21:51.000000000 +0100
Change: 2022-04-01 16:58:53.800000000 +0200
 Birth: -

TEST root@dennisosd2:~$ ls -al /usr/sbin/lvs
lrwxrwxrwx 1 root root 3 Feb 13  2020 /usr/sbin/lvs -> lvm
TEST root@dennisosd2:~$ ls -al /usr/sbin/pvs
lrwxrwxrwx 1 root root 3 Feb 13  2020 /usr/sbin/pvs -> lvm
```

But let not get distracted here.

If the assumption holds that "disk.lsblk" and "Device" should always return a correct (in the sense of the "true" state of rotational) value of key "ROTA" , than according to my tests both are broken: disk.lsblk in a hard way, while "Device" is silently broken (and returning a default "True" in case something goes wrong). I believe in case of LVM we end up here in the code: [rotational](https://github.com/guits/ceph/blob/dmcrypt-optim/src/ceph-volume/ceph_volume/util/device.py#L388) and a default of "1" gets returned.

So, how did I test and what?
1)  git clone your ceph repo
2 checkout the dmcrypt-optim branch
3) On a Ubuntu 20.04 VM running ceph 16.2.11-1 the "/usr/lib/python3/dist-packages/ceph_volume" directory was replaced with the ceph_volume directory in the dmcrypt-optim branch.
4) performed the tests

Just to make sure I'm not making any errors here: Can you reproduce the tests I did with LVM logical volume on your test system?

Reply:
@hydro-b ok, thanks for this clear and detailed report/testing 👍 
I'm gonna test the same in my environment to be sure we are not missing anything and try to shed the light on your findings.

Reply:
@hydro-b 

```console
# lsblk
NAME                                                                                                  MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINT
vda                                                                                                   253:0    0   11G  0 disk
`-vda1                                                                                                253:1    0   10G  0 part  /rootfs
vdb                                                                                                   253:16   0  200G  0 disk               #  <---- rotational device
`-vg_ceph2-lv_ceph2                                                                                   252:5    0  100G  0 lvm
vdc                                                                                                   253:32   0  200G  0 disk
vdd                                                                                                   253:48   0  200G  0 disk
vde                                                                                                   253:64   0  200G  0 disk
vdf                                                                                                   253:80   0  200G  0 disk
vdg                                                                                                   253:96   0  200G  0 disk
vdh                                                                                                   253:112  0  200G  0 disk
vdi                                                                                                   253:128  0  200G  0 disk
vdj                                                                                                   253:144  0  200G  0 disk
vdk                                                                                                   253:160  0  200G  0 disk
nvme0n1                                                                                               259:0    0  200G  0 disk
`-ceph--c673f4bb--c777--4ef9--b2fa--4ceff9848ea0-osd--block--210bd35f--841a--4e68--bef5--a65e56d9c539 252:0    0  200G  0 lvm
  `-OdCzrL-7smV-SuBf-Y6EE-ZmTg-R1zt-2smArW                                                            252:2    0  200G  0 crypt
nvme1n1                                                                                               259:1    0  200G  0 disk
`-ceph--50ab7485--e020--4763--83b5--36a61b19c5a7-osd--db--bf239e44--08f0--4d26--9fe8--affd21d49bf9    252:1    0  200G  0 lvm
  `-4eMOki-yIdt-BJJQ-qr4J-6dfy-KFuE-47bgMy                                                            252:3    0  200G  0 crypt
nvme2n1                                                                                               259:2    0  200G  0 disk               #  <---- non-rotational device
`-vg_ceph-lv_ceph                                                                                     252:4    0  100G  0 lvm
```


```console
# ls -l /dev/mapper/
total 0
lrwxrwxrwx. 1 ceph ceph       7 Nov  9 09:42 4eMOki-yIdt-BJJQ-qr4J-6dfy-KFuE-47bgMy -> ../dm-3
lrwxrwxrwx. 1 ceph ceph       7 Nov  9 09:42 OdCzrL-7smV-SuBf-Y6EE-ZmTg-R1zt-2smArW -> ../dm-2
lrwxrwxrwx. 1 root root       7 Nov  9 09:41 ceph--50ab7485--e020--4763--83b5--36a61b19c5a7-osd--db--bf239e44--08f0--4d26--9fe8--affd21d49bf9 -> ../dm-1
lrwxrwxrwx. 1 root root       7 Nov  9 09:41 ceph--c673f4bb--c777--4ef9--b2fa--4ceff9848ea0-osd--block--210bd35f--841a--4e68--bef5--a65e56d9c539 -> ../dm-0
crw-------. 1 root root 10, 236 Nov  9 08:47 control
lrwxrwxrwx. 1 root root       7 Nov 13 12:04 vg_ceph-lv_ceph -> ../dm-4
lrwxrwxrwx. 1 root root       7 Nov 13 12:10 vg_ceph2-lv_ceph2 -> ../dm-5
```

## rotational device

```console
# cat /sys/block/vdb/queue/rotational
1
# cat /sys/block/dm-5/queue/rotational
1
```

```console
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/vdb')['ROTA']
'1'
>>> disk.lsblk('/dev/vg_ceph2/lv_ceph2')['ROTA']
'1'
>>> disk.lsblk('/dev/dm-5')['ROTA']
'1'
>>> Device('/dev/vdb').rotational
True
>>> Device('/dev/vg_ceph2/lv_ceph2').rotational
device is not part of ceph: </dev/vg_ceph2/lv_ceph2>
True
>>> Device('/dev/dm-5').rotational
True
```


## Non-rotational device

```console
# cat /sys/block/nvme2n1/queue/rotational
0
# cat /sys/block/dm-4/queue/rotational
0
```

```console
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/nvme2n1')['ROTA']
'0'
>>> disk.lsblk('/dev/vg_ceph/lv_ceph')['ROTA']
'0'
>>> disk.lsblk('/dev/dm-4')['ROTA']
'0'
>>> Device('/dev/nvme2n1').rotational
False
>>> Device('/dev/vg_ceph/lv_ceph').rotational
device is not part of ceph: </dev/vg_ceph/lv_ceph>
False
>>> Device('/dev/dm-4').rotational
False
```

Reply:
@hydro-b 

You are running on `Ubuntu 20.04 VM` whereas I'm running on `CentOS Stream release 8`, hence all the log messages

```
Executable XXX not found on the host, will return lvs as-is` you are seeing.
```

As per LVM experts, LVM commands should never been run from within a container, so we had to introduce a behaviour in ceph-volume to make it run any LVM commands from the host using nsenter.

with Centos, LVM binaries are located in `/sbin`,
with Ubuntu, LVM binaries are located in `/usr/sbin`

I think the difference between our results is especially a consequence of this behaviour.
That being said, ceph-volume tries multiple paths on the host to find LVM binaries including `/usr/sbin` so it should work.
I'm going to deploy a new environment using Ubuntu and investigate that. I'll let you know ASAP.

Reply:
@hydro-b re-tested with Ubuntu focal and everything worked as expected, same results as with my previous test with CentOS 🤷‍♂️

Reply:
Thanks for testing again. I'm at a loss here ...

FWIW, on ubuntu there is a symlink for `/sbin`:
```
ls -al /sbin
lrwxrwxrwx 1 root root 8 Apr  1  2022 /sbin -> usr/sbin
```
So whatever ceph-volume would execute it should work.

I have a cephadm based Reef cluster on Ubuntu 20.04 (docker as container runtime). I modified a 18.2.0 container (`sha256:a2d68ce89c8910cde20540d5d17874a94987118b939fae2e31d48177496ea9a2` with your ceph-volume branch and replaced the ceph_volume directory.  I commited changes, tagged, pushed to private repository, pointed cephadm to new (global) image, and redeployed osd containers. That works. Re-creating an OSD with ceph original container image with encryption also works. If I redeploy this container with the modified ceph-volume image (while the underlying dm-crypt device is already active) it works as well. However, if I restart the container host the OSD cannot be brought online:

ceph-osd.1.log:
```
2023-11-14T10:12:59.567+0000 7fc503c4f640  0 set uid:gid to 167:167 (ceph:ceph)
2023-11-14T10:12:59.567+0000 7fc503c4f640  0 ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable), process ceph-osd, pid 7
2023-11-14T10:12:59.567+0000 7fc503c4f640  0 pidfile_write: ignore empty --pid-file
2023-11-14T10:12:59.567+0000 7fc503c4f640 -1 bluestore(/var/lib/ceph/osd/ceph-1/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-1/block: (2) No such file or directory
2023-11-14T10:12:59.567+0000 7fc503c4f640 -1 ESC[0;31m ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-1: (2) No such file or directoryESC[0m
2023-11-14T10:13:21.080+0000 7fd897df4640  0 set uid:gid to 167:167 (ceph:ceph
```

Ceph-volume logging:

```
[2023-11-14 11:33:11,764][ceph_volume.main][INFO  ] Running command: ceph-volume  activate --bad-option
[2023-11-14 11:33:13,485][ceph_volume.main][INFO  ] Running command: ceph-volume  activate --osd-id 1 --osd-uuid c2342c73-14c5-4c6a-a51a-94c1c273c2cf --no-systemd --no-tmpfs
[2023-11-14 11:33:13,487][ceph_volume.process][INFO  ] Running command: nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/
uts /usr/sbin/lvs --noheadings --readonly --separator=";" -a --units=b --nosuffix -S tags={ceph.osd_id=1,ceph.osd_fsid=c2342c73-14c5-4c6a-a51a-94c1c273c2cf} -o lv_tags,lv_path,lv_name,vg_nam
e,lv_uuid,lv_size
[2023-11-14 11:33:13,490][ceph_volume.process][INFO  ] stderr nsenter: failed to execute /usr/sbin/lvs: No such file or directory
[2023-11-14 11:33:14,743][ceph_volume.main][INFO  ] Running command: ceph-volume  lvm deactivate 1 c2342c73-14c5-4c6a-a51a-94c1c273c2cf
[2023-11-14 11:33:14,744][ceph_volume.process][INFO  ] Running command: nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/
uts /usr/sbin/lvs --noheadings --separator=";" -a -o lv_tags,lv_path,lv_name,vg_name,lv_uuid,lv_size -S lv_tags={ceph.osd_fsid=c2342c73-14c5-4c6a-a51a-94c1c273c2cf}
[2023-11-14 11:33:14,748][ceph_volume.process][INFO  ] stderr nsenter: failed to execute /usr/sbin/lvs: No such file or directory
[2023-11-14 11:33:14,748][ceph_volume.devices.lvm.deactivate][ERROR ] No data or block LV found for OSD1
```

So I decided to run some tests:

Running the ceph built 18.2.0 container image:

```
root@reef13:/root# docker run -it --privileged -v /dev:/dev -v /run/udev:/run/udev -v /sys:/sys -v /run/lvm:/run/lvm -v /run/lock/lvm:/run/lock/lvm -v /:/rootfs quay.io/ceph/ceph@sha256:a2d68ce89c8910cde20540d5d17874a94987118b939fae2e31d48177496ea9a2
[root@5534f72c86f3 /]# nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/uts /usr/sbin/lvs
nsenter: failed to execute /usr/sbin/lvs: No such file or directory

[root@5534f72c86f3 /]# lsblk 
NAME                                                                                                  MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda                                                                                                     8:0    0   20G  0 disk 
|-sda1                                                                                                  8:1    0    1M  0 part 
`-sda2                                                                                                  8:2    0   20G  0 part 
  |-LVM0-swap                                                                                         253:1    0    1G  0 lvm  [SWAP]
  `-LVM0-root                                                                                         253:2    0   19G  0 lvm  /rootfs
sdb                                                                                                     8:16   0  100G  0 disk 
`-ceph--9fe2ad66--a719--408b--ad5d--31c810be869c-osd--block--c2342c73--14c5--4c6a--a51a--94c1c273c2cf 253:3    0  100G  0 lvm  
sdc                                                                                                     8:32   0  100G  0 disk 
`-ceph--55cd063e--f236--4dc5--b625--899075150b9a-osd--block--380f35c9--41cd--4989--a2af--db3da262d839 253:4    0  100G  0 lvm  
sdd                                                                                                     8:48   0  100G  0 disk 
`-ceph--7e4efd3d--b53e--4103--98da--836fd486c046-osd--block--ff1d83ea--f50d--4c9d--bf68--2970045168d2 253:0    0  100G  0 lvm  
sr0                                                                                                    11:0    1  364K  0 rom  
[root@5534f72c86f3 /]# python3
Python 3.6.8 (default, Oct 25 2023, 14:42:08) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/sdb')['ROTA']
'0'
>>> 
[root@5534f72c86f3 /]# ls /dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf 
/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf

[root@5534f72c86f3 /]# python3
Python 3.6.8 (default, Oct 25 2023, 14:42:08) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf')['ROTA']
'0'
>>> Device('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf').rotational
True
```
Note that `disk.lsblk` is correct, but `Device().rotational` is not

Running the modified 18.2.0 container image with your branch:

```
root@reef13:/root# docker run -it --privileged -v /dev:/dev -v /run/udev:/run/udev -v /sys:/sys -v /run/lvm:/run/lvm -v /run/lock/lvm:/run/lock/lvm -v /:/rootfs reef.test.bit.nl:5000/guitsdmcrypt@sha256:aeb4b3ea16129b42688fa71164e83e4121c142d042dc047000b7c440ec64b07b

[root@81b292159e15 /]# nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/uts /usr/sbin/lvs
nsenter: failed to execute /usr/sbin/lvs: No such file or directory
[root@81b292159e15 /]# python3
Python 3.6.8 (default, Oct 25 2023, 14:42:08) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf')['ROTA']
'0'
>>> Device('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf').rotational
False
```
Note that both `disk.lsblk` and `Device().rotational` are correct.

So it seems a bug has been fixed between 18.2.0 and your branch (pretty much main), which is good news. However, I'm unable to get this running on my end. I might be overlooking something, not sure.

Would it be possible to build a container-image from your branch (ceph CI/CD)? I could try running that. Just to eliminate any user error I might be introducing in my tests.

Maybe @anthonyeleven and / or @baergj have time to test your branch and report their results. That might be interesting as well.

Reply:
@hydro-b I'll share here any outcome(s) anyway but it feels to me this discussion is becoming too long/verbose. I'm available on slack (ceph-storage.slack.com) / or IRC (OFTC) to direct message if you want...